Add useful functions/lemmas into the List library

jaybosamiya opened the pull request:
This PR introduces a bunch of useful things into the `List` library:

+ More properties for `splitAt`
+ `snoc`, `unsnoc` and their properites
+ `split_using` (splits at the first matching element it finds) and its properites
+ `split3` (splits into 3 pieces: left, element, right) and its properties
+ Properties connecting all of these together

At the moment, the CI seems to give me "success with breakages". Yet to analyze those breakages, so this should not be merged in yet, but I am making this PR so that review for it can be started, especially since it is modifying (albeit only via additions) into `ulib`.

nikswamy commented:
Thanks @jaybosamiya!

Some high-level comments here and more specific comments inline on the commits.

1. Can we get rid of List.Pure.Base and List.Pure.Base.Properties? @tahina-pro reminds me that the original intention of creating those files by @kyoDralliam was to circumvent some problems with bootstrapping. Is it still necessary? Is there another solution that doesn't involve creating another file in ulib?

2. If we're trying to go for an exemplary style, we should avoid naming lemmas with the `lemma_` prefix.

More specific comments within.

jaybosamiya commented:
It'd be _awesome_ to get rid of `List.Pure.Base` and `List.Pure.Base.Properties`, and merge into the main base and properties files, if we can. Perhaps that should be a separate PR though.

---

For the `lemma_` prefix, there are both pros and cons to keeping the prefix:

Pros:
+ Makes it easy during writing of proofs to quickly narrow down to lemmas by simply typing `lemma_` and then using auto-complete (Maybe this can be something added into `fstar-mode` to handle lemmas separately, making both code and proof writing easier?)
+ When reading a function that has both code and proofs in it, it is easy to quickly distinguish which part is code and which part is proof.

Cons:
+ (Unnecessary?) redundancy
+ Proofs end up having a whole line of `lemma_` one below another (which personally I don't mind, and in fact prefer, since it allows one to, at a glance, distinguish proof-related statements, and code-related statements).

I am okay with changing things over if need be, but I lean towards keeping the `lemma_` prefix due to the above mentioned pros. Whichever direction we decide, we should document into a wiki page.

---

On a separate note: should we have some kind of prefix for `SMTPat`d lemmas? One does not expect to call these directly, so by having a prefix, it reduces the effort of a proof-writer, since they don't have to worry about these at all. Something like `pat_` etc. If we follow such a style, then any invocations of a `pat_` function would mean that one is consciously breaking away from how the patterns are behaving.

---

I don't see any inline comments, so I just wanted to confirm if you've added any specific comments yet.

nikswamy commented on ulib/FStar.List.Pure.Properties.fst:
```diff
@@ -52,3 +52,200 @@ let rec splitAt_length_total (#a:Type) (l:list a)
+  match i with
+  | 0 -> ()
+  | _ -> lemma_splitAt_reindex_right (i - 1) (tl l) j
+
```
`splitAt n l` is unusual in that it's defined for all `n`. Was that intentional @kyoDralliam? Why not define it only for `n <= length l` or maybe even `n < length l`? That's a separate question independent of this PR.

Now, I was expecting to see the library provide a single lemma that characterizes the equational behavior of splitAt in terms of other existing notions for which we already have lemmas. For instance, a lemma like:

```
n <= length l ==>
(splitAt n l = l1, l2 <==>
l == l1 @ l2 /\
length l1 = n)
```

`lemma_splitAt_append` and `lemma_append_splitAt` come close to this, but do not quite constrain `n` sufficiently.

jaybosamiya replied:
I'm not sure I understand what the "`lemma_splitAt_append` and `lemma_append_splitAt` come close to this, but do not quite constrain `n` sufficiently" means. 

Since `n` is a `nat`, it is always `>=0`. Additionally, `lemma_splitAt_append` has a `requires` clause of `n <= length l`. I believe that this constraint (rather than `n < length l`) makes sense. Consider this example:

```
l = [ 0 ; 1 ; 2 ; 3 ]
//  ^   ^   ^   ^   ^
//  0   1   2   3   4
```

We would want all these possible points for splitting (i.e., `0 <= n <= length l`).

Also, should I merge the lemmas as per what you've said, or should I leave that for a separate PR?

nikswamy replied:
This is what I mean about "...not quite constrain `n` sufficiently"

```
--- a/ulib/FStar.List.Pure.Properties.fst
+++ b/ulib/FStar.List.Pure.Properties.fst
@@ -60,7 +60,7 @@ let rec lemma_splitAt_append (#a:Type) (n:nat) (l:list a) :
   Lemma
     (requires n <= length l)
     (ensures (let l1, l2 = splitAt n l in
-              append l1 l2 == l)) =
+              append l1 l2 == l /\ n = length l1)) =
   match n with
   | 0 -> ()
   | _ ->
@@ -78,6 +78,10 @@ let rec lemma_append_splitAt (#t:Type) (l1 l2:list t) :
   | [] -> ()
   | _ -> lemma_append_splitAt (tl l1) l2

+let rec splitAt_append (#t:Type) (l l1 l2:list t) (n:nat{n <= length l})
+  : Lemma (splitAt n l == (l1, l2) <==> l == l1 @ l2 /\ length l1 = n)
+  = lemma_splitAt_append n l;
+    lemma_append_splitAt l1 l2
```

I think a single `splitAt_append` lemma that fully characterizes the behavior of `splitAt n l` in terms of more standard list concepts like `@` and `length` would be useful. I've shown this at least for ` n <= length l`, but you could easily also prove a more general lemma (where if `n > length l` then `l2 = []`)

jaybosamiya replied:
This is already covered by the pre-existing `splitAt_length`, which is why I didn't include it here.

jaybosamiya replied:
Added this in.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -258,6 +268,13 @@ val rev_involutive : l:list 'a ->
+val snoc_length : l:list 'a -> x:'a ->
+  Lemma (requires True)
+        (ensures (length (snoc l x) = length l + 1))
+        [SMTPat (length (snoc l x))]
```
This smt pattern should be removed. `snoc` is defined in terms of `append` and `FStar.List.Tot.Properties.append_length` already has an SMT pattern.

jaybosamiya replied:
Makes sense, removing.

jaybosamiya replied:
Just realized now why I had kept it: For some reason, the pattern on `FStar.List.Tot.Properties.append_length` doesn't seem to want to trigger for `snoc` invocations. Not entirely sure why. I'm keeping it for now, until I can analyze it further.

nikswamy replied:
I suspect that it is firing, but that you may be in a setting where you are working with `fuel = 0` or something like that. Because, with 0 fuel, `length [x] = 1` is not provable with SMT. That is provable with the normalizer though.

jaybosamiya replied:
Removed it. When I pushed the curried version of `snoc`, I had to update some of the other lemmas to ensure that the checks pass. 

I believe we should add in a `length (fst (unsoc l)) = length l - 1` lemma though, with a pattern on `length (fst (unsnoc l))`. Previously, the places that used `unsnoc` depended on unsnoc being related over to snoc, which then gave it the length. Now, the unsnoc/snoc relation is not directly triggered via simply using unsnoc and length, and thus it makes anything with an unsoc'd list (for example, an `index` operation) much more cumbersome. What do you think?

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
+val lemma_unsnoc_snoc: #a:Type -> l:list a{length l > 0} ->
+  Lemma (requires True)
+    (ensures (let l', x = unsnoc l in snoc l' x == l))
+    [SMTPat (unsnoc l)]
```
This SMT pattern is worrisome. It fires on `unsnoc l` but contributes a term that contains `unsnoc l` as a strict sub term. This is risky as it may interact poorly with other patterns. Even aside from that, this pattern provides an equation about snoc/unsnoc in the context as soon as an unsnoc appears, which may be unnecessary. 

I suggest either

1. picking the pattern to be `snoc (fst (unsnoc l)) (snd (unsnoc l))`

2. removing the pattern and relying on explicit invocations at callsites.

jaybosamiya replied:
I'm going for the 1st option you suggested.

jaybosamiya replied:
From my testing, it looks like the pattern `snoc (fst (unsnoc l)) (snd (unsnoc l))` is too restrictive, and having an equation about snoc/unsnoc in context seems extremely useful (at least from my use cases, almost any time you want to use snoc, you also want unsnoc, and vice-versa).

I completely agree that purely `unsnoc l` is worrisome too, although I am not sure I understand the "as a strict subterm" part. Could you please clarify? Thanks :)

nikswamy replied:
I generally like to see patterns on quantifiers where the terms that result from the quantifier instantiation are smaller than the term used to instantiate the pattern. This is not a strict rule, but with this discipline, it is usually easier to argue that there are no matching loops, since the new terms produces after a quantifier instantiation is smaller than the pattern you started with.

nikswamy replied:
One possibility may be to define (uncurried) `snoc (l, x) = append l [x]` and then this lemma could be `snoc (unsnoc l) == l` with `snoc (unsnoc l)` as the pattern. I can see that `snoc (fst (unsnoc l)) (snd (unsnoc l))` could be too restrictive (since the `fst` and `snd` may not appear explcitly).

jaybosamiya replied:
Defining it with the uncurried version makes a lot of sense. In fact, I am pretty sure this is a better/cleaner definition. Should I also add in `lemma_snoc_unsnoc` (which says `unsnoc (snoc (l, x)) == (l, x)`) as well? I personally don't need it, but it seems like something that someone else using snoc/unsnoc might want/need.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
+(** [split_using] splits a list at the first instance of finding an
+    element in it. *)
+let rec split_using (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  GTot (r:(list t * list t)) =
```
In what cases do you need to split a list like this when `t` is not an `eqtype`?

jaybosamiya replied:
In my dll proof, I hold on to an `erased (list (pointer (node t)))`, and need to regularly split this given only a `pointer (node t)`. Since this is an `erased`, it is sufficient for me to be able to do this in the `GTot` world.

I'd be glad to hear alternative suggestions for this, but I believe it to be a useful thing to be able to do at the ghost-level.

nikswamy replied:
I can see how that might be useful.

I'm a bit sad to see StrongExcludedMiddle in the middle of the list library. But, I don't have an alternative to offer you without an in depth review and potential overhaul of your proof of a  doubly linked list. At a distance, if your `erased (list (pointer (node t)))` represents the ghost state of your linked list, then why does it contain pointers to nodes rather than the nodes themselves. Anyway, that's just a shot in the dark not knowing much about your proof.

At this stage, I am okay with accepting this function but it would be good to document it prominently saying that it uses this axiom internally.

jaybosamiya replied:
Yes, it does indeed represent the ghost state of the linked list. I had moved between `erased (list (pointer (node t)))` and `erased (list (node t))` multiple times in the process of building the proof, and from experience, I realized that holding on to the list of pointers helps to tremendously simplify the proof since a bunch of properties that are essential (almost counterintuitively) can be talked about without refering to heaps. By simplification, I mean both "understandability" of the proof, as well as the time the proof takes to verify successfully.

For now, I have added in a prominent `NOTE` in the doc string for this.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
+  match l with
+  | [_] -> ()
+  | a :: as ->
+    if FStar.StrongExcludedMiddle.strong_excluded_middle (a == x)
```
`strong_excluded_middle` is a very heavy hammer. I would suggest to use it sparingly. Here's a proof that doesn't require it:

```
let rec lemma_split_using (#t:Type) (l:list t) (x:t{x `memP` l}) :
  Lemma
    (ensures (
        let l1, l2 = split_using l x in
         length l2 > 0 /\
        ~(x `memP` l1) /\ 
         hd l2 == x /\
        append l1 l2 == l))
    [SMTPat (split_using l x)] =
  match l with
  | [_] -> ()
  | a :: as ->
    let goal = 
        let l1, l2 = split_using l x in
         length l2 > 0 /\
        ~(x `memP` l1) /\ 
         hd l2 == x /\
        append l1 l2 == l
    in
    FStar.Classical.or_elim 
      #_ #_ 
      #(fun () -> goal)
      (fun (_:squash (a == x)) -> ())
      (fun (_:squash (x `memP` as)) -> lemma_split_using as x)
```

jaybosamiya replied:
Huh, I didn't realize I could do this this way. Thanks a lot! The `strong_excluded_middle` definitely felt like a heavy hammer to use, but I wasn't sure if it could be done any other way. Will update based upon this.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
+
+(** [index_of l x] gives the index of the leftmost [x] in [l] *)
+let rec index_of (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  GTot (i:nat{i < length l /\ index l i == x}) =
```
Same question as above: when do you need this for a non-`eqtype`?

jaybosamiya replied:
Similar answer as earlier, I sometimes need to (in a ghostly sense) get the index of an element which I am holding on to, but it is not an `eqtype` since it is a `pointer`.

nikswamy replied:
same as above. Document prominently mentioning the axiom it uses.

jaybosamiya replied:
documented.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -426,6 +538,15 @@ let rec mem_memP
+let rec lemma_index_memP (#t:Type) (l:list t) (i:nat{i < length l}) :
+  Lemma
+    (ensures (index l i `memP` l))
+    [SMTPat (index l i `memP` l)] =
```
Ok. This pattern looks sufficiently restrictive to me.

nikswamy commented on ulib/FStar.List.Tot.Properties.fst:
```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
+        (length l2 > 0) /\
+        ~(x `memP` l1) /\ (hd l2 == x) /\
+        append l1 l2 == l))
+    [SMTPat (split_using l x)] =
```
This destructs every occurrence of a `split_using` into other predicates. Is it necessary to do that eagerly on every occurrence of `split_using`? It would be safer to remove the pattern and let the user call this lemma as needed.

jaybosamiya replied:
I guess this makes sense. I will remove this pattern.

nikswamy commented:
There are many debatable points of style and convention. Some of them are shallow: e.g., `lemma_` or not. Others are deeper: should we have so many specialized lemmas about derived forms in the library. It's going to be hard to resolve these questions at this stage.

In general, I'm in favor of accepting this PR after the specific questions I raised are addressed. Then after merging, let's take this as an opportunity to review the general design principles of the standard library as a whole, starting with lists.

jaybosamiya commented:
We seem to have resolved all the comments, except for minor lingering questions:
- [ ] https://github.com/FStarLang/FStar/pull/1560#discussion_r226037931
  - (Should we add in a lemma with a pattern on `length (fst (unsoc l))`? This becomes very necessary since we've restricted the `lemma_unsnoc_snoc` pattern)
- [ ] https://github.com/FStarLang/FStar/pull/1560#discussion_r226035584
  - (Should we add in a `unsnoc (snoc (l, x)) == (l, x))` lemma as well? This seems like something people might want out of the properties of snoc/unsnoc)

nikswamy commented:
Regarding the first point: I think including (fst (unsnoc l)) in the pattern is likely to restrictive. The first component of a pair is not always projected using `fst` so this may not always trigger as expected. It's what you observed on one of my suggestions for snoc/unsnoc. I don't have a great alternative however. Maybe, in this case, `length (Mktuple2?._1 (unsnoc l))` will be better. But then again isn't this already covered by one of the lemmas about splitAt?

Yes on the second one too. But I think the lemma can just be stated as `unsnoc (snoc l) == l` rather than destructing the pair into its components needlessly.

A third remaining point: before merging, especially since this introduces new smt patterns in a core library, I'd like to see this branch get a Everest green before merging. @protz, @aseemr ...can one of you help Jay trigger an appropriate Everest build? (I'm on my phone)

Thanks for all the rounds of work, Jay!
________________________________
From: Jay Bosamiya <notifications@github.com>
Sent: Wednesday, October 17, 2018 11:42:09 AM
To: FStarLang/FStar
Cc: Nikhil Swamy; Comment
Subject: Re: [FStarLang/FStar] Add useful functions/lemmas into the List library (#1560)


We seem to have resolved all the comments, except for minor lingering questions:

  *    #1560 (comment)<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FFStarLang%2FFStar%2Fpull%2F1560%23discussion_r226037931&data=02%7C01%7Cnswamy%40microsoft.com%7Cc915b9c5f86342ba328d08d634603ec9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636753985307315407&sdata=JMmeEqmM8AthQ4DDGnnPirxlzZM9ZP54mYZW%2F2EDd5Y%3D&reserved=0>
     *   (Should we add in a lemma with a pattern on length (fst (unsoc l))? This becomes very necessary since we've restricted the lemma_unsnoc_snoc pattern)
  *    #1560 (comment)<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FFStarLang%2FFStar%2Fpull%2F1560%23discussion_r226035584&data=02%7C01%7Cnswamy%40microsoft.com%7Cc915b9c5f86342ba328d08d634603ec9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636753985307315407&sdata=WkrXuPHTxNFoiSTUILHDFUbvjzCZYFwW%2BcyDB7FmEU8%3D&reserved=0>
     *   (Should we add in a unsnoc (snoc (l, x)) == (l, x)) lemma as well? This seems like something people might want out of the properties of snoc/unsnoc)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FFStarLang%2FFStar%2Fpull%2F1560%23issuecomment-430742783&data=02%7C01%7Cnswamy%40microsoft.com%7Cc915b9c5f86342ba328d08d634603ec9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636753985307325416&sdata=y%2FhFE4KNJ1C4jQFd8IhI0QE7vBfY6jCBzFcD7Qg7pGc%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAF4Ln2Ncowb40hpDwGaiTsjl-NYa33AZks5ul3oBgaJpZM4XUD6i&data=02%7C01%7Cnswamy%40microsoft.com%7Cc915b9c5f86342ba328d08d634603ec9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636753985307325416&sdata=SeXvf5DW1RiPRGUKqfOQL6xdh8FxdJqcBfIXJ3L0bHI%3D&reserved=0>.

jaybosamiya commented:
Ah, the `Mktuple2?._1` thing makes sense. Added it in. 

Unfortunately, this still tends to be too restrictive. Consider the following example:

```
let rec test #t (l: list t) :
  Pure unit (requires True) (ensures (fun _ -> True)) (decreases (length l)) =
  match length l with
  | 0 -> ()
  | _ ->
    let xs, x = unsnoc l in
    lemma_unsnoc_length l; // cannot prove termination without this
    test xs
```

This seems like a common enough use case (i.e. recursing backwards on a list). Should we instead just place a refinement on `unsnoc`'s result itself? Recursions that use `unsnoc` are _always_ going to need to call out to this lemma to prove termination otherwise.

The `splitAt` lemmas don't have patterns on them, and thus we cannot directly use them to get around this either.

---

Added in the lemma for the `unsnoc (snoc l) == l`

---

For the everest build, that makes sense. Let's trigger it after we have made a decision on `lemma_unsnoc_length`.

---

Thanks a lot for your detailed comments, Nik! It made it much easier to understand how to improve this PR to a better state :)

nikswamy commented:
If we don't have a robust pattern for unsnoc_length then I'd prefer no pattern at all. It's more predictable that way.
________________________________
From: Jay Bosamiya <notifications@github.com>
Sent: Wednesday, October 17, 2018 1:04:39 PM
To: FStarLang/FStar
Cc: Nikhil Swamy; Comment
Subject: Re: [FStarLang/FStar] Add useful functions/lemmas into the List library (#1560)


Ah, the Mktuple2?._1 thing makes sense. Added it in.

Unfortunately, this still tends to be too restrictive. Consider the following example:

let rec test #t (l: list t) :
  Pure unit (requires True) (ensures (fun _ -> True)) (decreases (length l)) =
  match length l with
  | 0 -> ()
  | _ ->
    let xs, x = unsnoc l in
    lemma_unsnoc_length l; // cannot prove termination without this
    test xs


This seems like a common enough use case (i.e. recursing backwards on a list). Should we instead just place a refinement on unsnoc's result itself? Recursions that use unsnoc are always going to need to call out to this lemma to prove termination otherwise.

The splitAt lemmas don't have patterns on them, and thus we cannot directly use them to get around this either.

________________________________

Added in the lemma for the unsnoc (snoc l) == l

________________________________

For the everest build, that makes sense. Let's trigger it after we have made a decision on lemma_unsnoc_length.

________________________________

Thanks a lot for your detailed comments, Nik! It made it much easier to understand how to improve this PR to a better state :)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FFStarLang%2FFStar%2Fpull%2F1560%23issuecomment-430768973&data=02%7C01%7Cnswamy%40microsoft.com%7C80b12102bacf4084356b08d6346bc5bd%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636754034817345354&sdata=%2F1wc1Vwn4BcOrMRPec9yCKSh2vaAVfJZhdsqp2NDgi8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAF4Ln4EOnCRzQR4cACI79ylkzNpSrCYxks5ul41XgaJpZM4XUD6i&data=02%7C01%7Cnswamy%40microsoft.com%7C80b12102bacf4084356b08d6346bc5bd%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636754034817345354&sdata=ATeRk9hIaAjMi9jdligC0zgzYqwrDTazUHHNAq5hop0%3D&reserved=0>.

jaybosamiya commented:
Sounds good, removed the pattern.

msprotz commented:
Pushed as list_sn for Everest. If all goes well, a notification should be issue in the Slack #everest-build channel

jaybosamiya commented:
Thanks a lot Jonathan! :)

jaybosamiya commented:
Got the green light to merge from Jonathan. Checked the logs to ensure that the right version was used on the Everest green (just a sanity check). Merged.

Thanks a lot @nikswamy and @msprotz !! :smiley:
